finality-aleph/src/runtime_api.rs

```diff
@@ -63,36 +64,74 @@ where
         }
     }
 
-    fn read_storage<D: Decode>(
+    fn access_storage<D: Decode>(
+        &self,
+        storage_key: Vec<u8>,
+        at_block: BlockHash,
+    ) -> Result<D, ApiError> {
+        let encoded = self
+            .client
+            .storage(at_block, &sc_client_api::StorageKey(storage_key))
+            .map_err(|_| ApiError::StorageAccessFailure)?
+            .ok_or(ApiError::NoStorage)?;
+        D::decode_all(&mut encoded.0.as_ref()).map_err(ApiError::DecodeError)
+    }
+
+    fn read_storage_value<D: Decode>(
         &self,
         pallet: &str,
         item: &str,
         at_block: BlockHash,
     ) -> Result<D, ApiError> {
         let storage_key = [twox_128(pallet.as_bytes()), twox_128(item.as_bytes())].concat();
+        match self.access_storage(storage_key, at_block) {
+            Err(ApiError::NoStorage) => Err(ApiError::NoStorageValue(pallet.into(), item.into())),
+            other => other,
+        }
+    }
 
-        let encoded = match self
-            .client
-            .storage(at_block, &sc_client_api::StorageKey(storage_key))
-        {
-            Ok(Some(e)) => e,
-            _ => return Err(ApiError::NoStorage(pallet.to_string(), item.to_string())),
-        };
-
-        D::decode_all(&mut encoded.0.as_ref()).map_err(ApiError::DecodeError)
+    #[allow(dead_code)]
+    fn read_storage_map<H: StorageHasher, D: Decode, E: Encode>(
```

Comment:
Is this method supposed to read key value from only from `StorageMap` which keys are hashed using `Blake2_128Concat`? As there are `StorageMap`s like `Staking.bonded` that use `Twox64Concat`

Reply:
Ah no, all good - it's first template parameter!